File: Dynamic Algorithm/c++/canSumTabulation.cpp
#include<bits/stdc++.h>
using namespace std;

bool canSum(int target, vector<int>numbers)
{
    vector<bool>memo;
    memo.assign(target+1,false);
    memo[0]=true;
    int n_numbers=numbers.size();

    for (int i=0;i<=target;i++)
        if (memo[i])
        {
            for (int j=0;j<n_numbers;j++)
                if (numbers[j]+i<=target)
                    memo[i+numbers[j]]=true;
        }

    return memo[target];
}

int main()
{
    cout<<canSum(7,{2,3})<<endl;
    cout<<canSum(7,{15, 3, 4, 7})<<endl;
    cout<<canSum(7,{2,4})<<endl;
}

//https://www.youtube.com/watch?v=oBt53YbR9Kk&t=13079s

File: Dynamic Algorithm/python/canSumTabulation.py
from cmath import pi

def canSum(target, numbers):
    memo=[False for i in range(target+1)]
    memo[0]=True

    for i in range(target+1):
        if memo[i]:
            for num in numbers:
                if i+num<=target:
                    memo[i+num]=True

    return memo[target]

print(canSum(7, [2, 3]))
print(canSum(7, [15, 3, 4, 7]))
print(canSum(7, [2, 4]))
print(canSum(8, [2, 3, 5]))
print(canSum(300, [7, 14]))